File: src/lib/api/rcPair.ts
/**
 * 远程电脑 · 局域网 6 位数字配对（A3）—— 对应 `src-tauri/src/commands/rc_pair.rs`。
 *
 * 与 `lib/api/rc.ts` **分开**：那边是邀请码 / 会话 / 目标列表（700+ 行的命令层），
 * 这边是**独立的一条路**——不走邀请码、不走 Endpoint 握手，靠 presence 上的
 * 明文包 + 6 位数字核对完成。后端同样把命令层拆成两个文件，这里对齐。
 *
 * # 与邀请码那条路的关系
 *
 * 邀请码是**兜底**：不在同一局域网时只能靠带外渠道搬一串码。局域网这条路把
 * 「搬字符串」整个消掉，而且**信任强度更高**（中间人换公钥会让两端数字对不上）。
 * 两条路配完之后写的是同一张 `rc_devices`，界面上没有区别。
 *
 * 字段名与 Rust 一致（snake_case）。失败由调用方处理，这里不吞。
 */
import { invoke } from "@tauri-apps/api/core";

/** 同网段里**还没配对**的邻居（已配对的在 `rcTargets` 里，不在这一份）。 */
export interface RcNeighbor {
  node_id: string;
  /** 对方自报的名字，可能为空串。**不可信**，界面要与指纹一起标注。 */
  name: string;
  /** 对端自报的可达地址 `ip:port`。仅诊断用，界面不展示。 */
  addr: string;
  /** 本机最后一次听到它的时刻（本机时钟，epoch 毫秒）。 */
  last_seen_ms: number;
}

/** 正在进行的那一轮配对。 */
export interface RcPairPrompt {
  peer_id: string;
  peer_name: string;
  /** 6 位数字；空串 = 还在等对方的公钥（界面显示「正在与对方核对…」）。 */
  pin: string;
  /** 本端是不是发起方。只影响文案。 */
  initiator: boolean;
  /** 本端已点确认。 */
  me_ok: boolean;
  /** 对方已回「我这侧也确认了」。 */
  peer_ok: boolean;
  started_ms: number;
}

/** 刚配对成功的那一台（后端**读完即清**，前端要留在自己状态里）。 */
export interface RcPairDone {
  peer_id: string;
  peer_name: string;
  /** 本端是不是发起方 —— 决定完成屏给「立刻发起远程」还是「知道了」。 */
  initiator: boolean;
  at_ms: number;
}

export interface RcNearbyStatus {
  neighbors: RcNeighbor[];
  pair: RcPairPrompt | null;
  done: RcPairDone | null;
}

export interface RcPairOutcome {
  /** waiting = 本端确认了、等对方；committed = 两端都确认、**已配对**；gone = 会话已过期。 */
  state: "waiting" | "committed" | "gone";
  peer_id: string;
  peer_name: string;
}

/**
 * 附近设备 + 当前配对状态。
 *
 * ❗ 后端在这条命令里**顺带重传**该重传的握手包（见 `commands/rc_pair.rs`），
 * 所以界面对话框开着时要保持 2 秒一轮——关掉轮询等于放弃 UDP 丢包的重传机会。
 */
export function rcNearbyStatus(): Promise<RcNearbyStatus> {
  return invoke<RcNearbyStatus>("rc_nearby_status");
}

export function rcNearbyPair(peerId: string): Promise<RcPairPrompt> {
  return invoke<RcPairPrompt>("rc_nearby_pair", { peerId });
}

export function rcNearbyConfirm(): Promise<RcPairOutcome> {
  return invoke<RcPairOutcome>("rc_nearby_confirm");
}

/** 返回是否记了一次「拒绝」（本端是被请求方时才记）。 */
export function rcNearbyCancel(): Promise<boolean> {
  return invoke<boolean>("rc_nearby_cancel");
}

File: src-tauri/src/commands/rc_pair.rs
//! 远程电脑 · 局域网配对的命令层（A3，2026-09-17）。
//!
//! 与 `commands/rc.rs` **分开**：那个文件 700+ 行（邀请码、会话、目标列表、
//! 邀请门都在里面），而局域网配对是**独立的一条路**——不走邀请码、不走 Endpoint
//! 握手，靠 presence 上的明文包 + 6 位数字核对完成。混进那个文件只会让
//! 「改哪一段」变成靠记忆的事。
//!
//! # 与邀请码那条路的关系
//!
//! 邀请码（`commands/rc.rs::rc_invite_create` / `rc_pair`）是**兜底**：
//! 不在同一局域网时只能靠带外渠道搬一串码。这条局域网路径把「搬字符串」
//! 整个消掉，而且**信任强度更高**（中间人换公钥会让两端数字对不上，
//! 见 `rc/pin.rs` 的模块头）。
//!
//! 两条路的产物是同一张表（`rc_devices`），配完之后界面上没有区别。

use crate::rc::pin::{Confirmed, Done, PairPrompt};
use crate::rc::service::RcService;
use crate::sync::presence::Neighbor;
use serde::Serialize;
use std::sync::Arc;
use tauri::State;

/// 附近设备 + 当前配对状态。
#[derive(Debug, Clone, Serialize)]
pub struct RcNearbyStatus {
    /// 同一局域网里**还没配对**的邻居，最近听到的排前面。
    ///
    /// ❗ 名字是**对方自报**的，界面必须与指纹一起显示并标注「可自称」——
    /// 见 `wire::clean_name` 的注释。
    pub neighbors: Vec<Neighbor>,
    /// 正在进行的那一轮配对（没有则为 `null`）。
    pub pair: Option<PairPrompt>,
    /// 刚配对成功的那一台。
    ///
    /// 🔴 P1-7（2026-09-23 审计）：语义从「读完即清」改成「**60 秒窗口内多重可读**」——
    /// 主窗口与工作台两个轮询者并发调本命令，旧语义下先读到的那个把成功屏独占走，
    /// 另一个永远看不见。载荷形状没变（`Done`，带 `at_ms`），所以「这一条我已经弹过了」
    /// 的去重归界面（按 `at_ms` 比对），不归后端。
    pub done: Option<Done>,
}

/// 本端点了「确认」之后的结局。
#[derive(Debug, Clone, Serialize)]
pub struct RcPairResult {
    /// `waiting` = 本端确认了、在等对方；`committed` = 两边都确认了、**已配对**；
    /// `gone` = 会话已过期或被取消。
    pub state: String,
    pub peer_id: String,
    pub peer_name: String,
}

fn now_ms() -> i64 {
    chrono::Utc::now().timestamp_millis()
}

/// 附近设备 + 当前配对状态。
///
/// 🔴 界面对话框打开期间按 **2 秒**轮询它，而这个命令**顺带把该重传的配对包
/// 重传掉**（`RcService::nearby_prompt` → `discovery.tick`）。这是刻意的：
/// 配对只有 60 秒窗口，UDP 丢一个包就卡住，而「有人在看着」恰好是重传的前提
/// ——没有界面在读状态时，也就没有人在等这次配对。
#[tauri::command]
pub fn rc_nearby_status(svc: State<'_, Arc<RcService>>) -> RcNearbyStatus {
    let now = now_ms();
    RcNearbyStatus {
        neighbors: svc.nearby_neighbors(now),
        pair: svc.nearby_prompt(now),
        done: svc.nearby_done(now),
    }
}

/// 对着一台附近的设备发起配对。
///
/// 返回的 `pair.pin` 多半是空串——在等对方的公钥（一个来回，局域网内毫秒级）。
/// 界面这时该显示「正在与对方核对…」而不是一个空框。
#[tauri::command]
pub fn rc_nearby_pair(
    peer_id: String,
    svc: State<'_, Arc<RcService>>,
) -> Result<PairPrompt, String> {
    svc.nearby_pair_start(&peer_id, now_ms())
}

/// 本端点了「两边一样，确认」。
///
/// `waiting` 不是错误：两端的确认有先后，先点的那一侧本来就要等。
#[tauri::command]
pub fn rc_nearby_confirm(svc: State<'_, Arc<RcService>>) -> Result<RcPairResult, String> {
    let c = svc.nearby_confirm(now_ms())?;
    Ok(match c {
        Confirmed::Waiting { peer_id } => RcPairResult {
            state: "waiting".to_string(),
            peer_id,
            peer_name: String::new(),
        },
        Confirmed::Committed { peer_id, peer_name } => RcPairResult {
            state: "committed".to_string(),
            peer_id,
            peer_name,
        },
        Confirmed::Gone => RcPairResult {
            state: "gone".to_string(),
            peer_id: String::new(),
            peer_name: String::new(),
        },
    })
}

/// 取消这一轮配对。
///
/// 本端是**被请求**的那一侧时，顺带记一次「拒绝」——30 分钟内对方再来敲门
/// 不会再弹窗（`RcJoins::deny`）。发起方取消不记：那只是「我不发了」。
#[tauri::command]
pub fn rc_nearby_cancel(svc: State<'_, Arc<RcService>>) -> bool {
    svc.nearby_cancel(now_ms())
}
